packages/devtools/src/sourceMap/loadOriginalSourceLoc.ts

```diff
@@ -0,0 +1,178 @@
+  }
+
+  // No inline sourcemap found. Check if there's external source map
+  let sourceMappingURL = externalSourceMapURLs.pop();
+  if (!sourceMappingURL) {
+    throw new Error(
+      `[Griffel devtools] extractAndLoadSourceMapJSON() encountered string "sourceMappingURL" in file ${source} that does not contain a valid source map`,
+    );
+  }
+
+  if (externalSourceMapURLs.length > 0) {
+    console.warn(
+      `[Griffel devtools] extractAndLoadSourceMapJSON() More than one external source map detected in the file ${source}:`,
+    );
+    externalSourceMapURLs.forEach(sourceMappingURL => {
+      console.warn(`[Griffel devtools] extractAndLoadSourceMapJSON() skipping source map "${sourceMappingURL}"`);
+    });
+  }
+
+  if (!sourceMappingURL.startsWith('http') && !sourceMappingURL.startsWith('/')) {
+    const lastSlashIdx = source.lastIndexOf('/');
+    if (lastSlashIdx !== -1) {
+      const baseURL = source.slice(0, lastSlashIdx);
+      sourceMappingURL = `${baseURL}/${sourceMappingURL}`;
+    }
+  }
+
+  const sourceMapContent = await fetchFiles(sourceMappingURL);
+  const sourceMapJSON = JSON.parse(sourceMapContent);
+  sourceMapJSONs.set(source, sourceMapJSON);
+
+  return getOriginalPosition(sourceMapJSON, sourceLoc);
+}
+
+async function fetchFiles(url: string): Promise<string> {
+  try {
+    const response = await fetch(url);
+    if (response.ok) {
+      return response.text();
+    }
+    throw new Error(`[Griffel devtools] fetchRuntimeSource() bad response fetching ${url}: ${response.status}`);
+  } catch (error) {
+    // try to get url content from chrome resources api
+    const resource = (await resources).find(resource => resource.url === url);
+    if (resource) {
+      return new Promise(resolve => resource.getContent(content => resolve(content)));
+    }
+    throw new Error(`[Griffel devtools] fetchFiles() unable to fetch ${url}: ${error}`);
+  }
+}
+
+function decodeBase64String(encoded: string): string {
+  if (typeof atob === 'function') {
+    return atob(encoded);
+  } else if (typeof Buffer !== 'undefined' && Buffer !== null && typeof Buffer.from === 'function') {
+    return Buffer.from(encoded, 'base64').toString('utf-8');
+  } else {
+    throw Error('[Griffel devtools] Cannot decode base64 string');
+  }
+}
+
+type IndexSourceMapSection = {
+  map: IndexSourceMap | RawSourceMap;
+  offset: {
+    line: number;
+    column: number;
+  };
+};
+type IndexSourceMap = {
+  file?: string;
+  mappings?: void;
+  sourcesContent?: void;
+  sections: IndexSourceMapSection[];
+  version: number;
+};
+function sourceMapIncludesSource(sourcemap: RawSourceMap | IndexSourceMap, sourcePath: string): boolean {
+  if (sourcemap.mappings === undefined) {
+    return (sourcemap as IndexSourceMap).sections.some(section => {
+      return sourceMapIncludesSource(section.map, sourcePath);
+    });
+  }
+
+  return (sourcemap as RawSourceMap).sources.some(s => s === 'Inline Babel script' || s.includes(sourcePath));
+}
```

Comment:
```suggestion
function sourceMapIncludesSource(sourcemap: RawSourceMap | IndexSourceMap, sourcePath: string): boolean {
  const sourcemaps: (IndexSourceMap | RawSourceMap)[] = [sourcemap];
  while(sourcemaps.length) {
    const curSourcemap = sourcemaps.pop()!;
    if (isRawSourceMap(curSourcemap)) {
      return curSourcemap.sources.some(s => s === 'Inline Babel script' || s.includes(sourcePath));
    }

    sourcemaps.push(...curSourcemap.sections.map(section => section.map));
  }

  return false;
}

function isRawSourceMap(sourcemap: RawSourceMap | IndexSourceMap): sourcemap is RawSourceMap {
  return !!sourcemap.mappings;
}
```
is perf an issue ? I see you've patched quick-sort, so if it is the iterative dive should be cheaper in theory. Also you can avoid the as casing with a simple type predicate

Reply:
Perf is not an issue. I patched quick-sort because it uses `eval` and it is banned from devtools. I would prefer the recursive version because it's more readable IMO.

Reply:
In that case I would suggest you still use the type predicate to avoid `as` casting

Reply:
Updated with `isRawSourceMap` instead of using `as`, it looks much better IMO :)